File: rust/src/utils/test_runner.rs
// #exonware/xwsystem/rust/src/utils/test_runner.rs
//! #exonware/xwsystem/src/exonware/xwsystem/utils/test_runner.py
//! 
//! Reusable pytest runner utilities for all eXonware libraries.
//! 
//! Implements the hierarchical runner utilities described in:
//! - docs/guides/GUIDE_DEV.md
//! - docs/guides/GUIDE_TEST.md
//! 
//! Company: eXonware.com
//! Author: eXonware Backend Team
//! Email: connect@exonware.com
//! Version: 0.1.0.1
//! Generation Date: 28-Dec-2025


use std::collections::HashMap;

use crate::__future__::{annotations};
use std::path::{Path};

use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::fs;
use std::io::Write;

/// Capture output for terminal and Markdown file simultaneously.
pub struct DualOutput {
    pub output_file: PathBuf,
    _markdown_lines: Vec<String>,
}

impl DualOutput {
    /// Constructor
    pub fn new(output_file: PathBuf) -> Self {
        Self {
            output_file,
            _markdown_lines: Vec::new(),
        }
    }

    pub fn print(&mut self, text: &str, markdown_format: Option<&str>, emoji: Option<&str>) {
        let prefix = emoji.map(|e| format!("{} ", e)).unwrap_or_default();
        let line = format!("{}{}", prefix, text);
        println!("{}", line);
        self._markdown_lines.push(
            markdown_format.unwrap_or(&line).to_string()
        );
    }

    pub fn save(&self, header_info: Option<&HashMap<String, String>>) -> Result<(), std::io::Error> {
        let header_info = header_info.unwrap_or(&HashMap::new());
        use std::time::{SystemTime, UNIX_EPOCH};
        let now = SystemTime::now()
            .duration_since(UNIX_EPOCH)
            .unwrap()
            .as_secs();
        let now_str = format!("{}", now); // Simple timestamp
        let header = format!(
            "# Test Runner Output\n\n\
            **Library:** {}  \n\
            **Layer:** {}  \n\
            **Generated:** {}  \n\
            **Runner:** {}  \n\n\
            ---\n\n",
            header_info.get("library").unwrap_or(&String::new()),
            header_info.get("layer").unwrap_or(&String::new()),
            now_str,
            header_info.get("runner").unwrap_or(&"TestRunner".to_string())
        );
        
        let mut file = fs::File::create(&self.output_file)?;
        file.write_all(header.as_bytes())?;
        file.write_all(self._markdown_lines.join("\n").as_bytes())?;
        file.write_all(b"\n")?;
        Ok(())
    }
}

#[derive(Debug, Clone)]
pub struct TestRunResult {
    pub exit_code: i32,
}

impl TestRunResult {
    pub fn new(exit_code: i32) -> Self {
        Self { exit_code }
    }
}

/// Simple reusable test runner for a single directory/layer.
pub struct TestRunner {
    pub library_name: String,
    pub layer_name: String,
    pub description: String,
    pub test_dir: PathBuf,
    pub markers: Vec<String>,
    pub output_file: PathBuf,
    pub output: DualOutput,
}

impl TestRunner {
    /// Constructor
    pub fn new(
        library_name: String,
        layer_name: String,
        description: String,
        test_dir: PathBuf,
        markers: Option<Vec<String>>,
        output_file: Option<PathBuf>,
    ) -> Self {
        let output_file = output_file.unwrap_or_else(|| test_dir.join("runner_out.md"));
        Self {
            library_name,
            layer_name,
            description,
            test_dir,
            markers: markers.unwrap_or_default(),
            output_file: output_file.clone(),
            output: DualOutput::new(output_file),
        }
    }

    pub fn run(&mut self) -> i32 {
        _configure_windows_utf8();

        print_header(&self.description, Some(&mut self.output));
        self.output.print(
            &format!("Directory: {}", format_path(&self.test_dir)),
            Some(&format!("**Directory:** `{}`", format_path(&self.test_dir))),
            Some("📂"),
        );

        let result = run_cargo_test(&self.test_dir, &self.markers);
        let ok = result.exit_code == 0;
        print_status(ok, if ok { "PASSED" } else { "FAILED" }, Some(&mut self.output));

        let mut header_info = HashMap::new();
        header_info.insert("library".to_string(), self.library_name.clone());
        header_info.insert("layer".to_string(), self.layer_name.clone());
        header_info.insert("runner".to_string(), "TestRunner".to_string());
        
        let _ = self.output.save(Some(&header_info));
        result.exit_code
    }
}

/// Configure UTF-8 for Windows consoles (GUIDE_TEST.md compliance).
pub fn _configure_windows_utf8() {
    #[cfg(windows)]
    {
        // On Windows, Rust already handles UTF-8 properly in most cases
        // This is a placeholder for any Windows-specific UTF-8 configuration
    }
}

/// Format a path as a full absolute path string.
pub fn format_path(path: &Path) -> String {
    path.canonicalize()
        .unwrap_or_else(|_| path.to_path_buf())
        .to_string_lossy()
        .to_string()
}

pub fn print_header(title: &str, output: Option<&mut DualOutput>) {
    let sep = "=".repeat(80);
    if let Some(out) = output {
        out.print(&sep, None, None);
        out.print(title, None, None);
        out.print(&sep, None, None);
    } else {
        println!("{}", sep);
        println!("{}", title);
        println!("{}", sep);
    }
}

pub fn print_section(title: &str, output: Option<&mut DualOutput>) {
    if let Some(out) = output {
        out.print(&format!("\n{}", title), Some(&format!("\n## {}\n", title)), None);
    } else {
        println!("\n{}\n{}\n{}\n", "=".repeat(80), title, "=".repeat(80));
    }
}

pub fn print_status(success: bool, message: &str, output: Option<&mut DualOutput>) {
    let emoji = if success { "✅" } else { "❌" };
    if let Some(out) = output {
        out.print(message, Some(&format!("**Result:** {} {}", emoji, message)), Some(emoji));
    } else {
        println!("{} {}", emoji, message);
    }
}

/// Run cargo test as a subprocess with standard eXonware flags.
pub fn run_cargo_test(test_dir: &Path, _markers: &[String]) -> TestRunResult {
    use std::process::Command;
    
    let result = Command::new("cargo")
        .arg("test")
        .arg("--manifest-path")
        .arg(test_dir.join("Cargo.toml"))
        .arg("--")
        .arg("--nocapture")
        .current_dir(test_dir)
        .status();
    
    match result {
        Ok(status) => TestRunResult::new(status.code().unwrap_or(1)),
        Err(_) => TestRunResult::new(1),
    }
}

File: src/exonware/xwsystem/utils/test_runner.py
#!/usr/bin/env python3
"""
#exonware/xwsystem/src/exonware/xwsystem/utils/test_runner.py
Reusable pytest runner utilities for all eXonware libraries.
Implements the hierarchical runner utilities described in:
- docs/guides/GUIDE_DEV.md
- docs/guides/GUIDE_TEST.md
Company: eXonware.com
Author: eXonware Backend Team
Email: connect@exonware.com
Version: 0.9.0.34
Generation Date: 28-Dec-2025
"""

from __future__ import annotations
import sys
import subprocess
from dataclasses import dataclass
from datetime import datetime
from pathlib import Path
from typing import Any
from exonware.xwsystem.console.cli import ensure_utf8_console


def timestamp_for_filename() -> str:
    """
    Return timestamp for time-sensitive log/test/review file names.
    Format: YYYYMMDD_HHMMSS_mmm (hour, minute, second, millisecond).
    Per GUIDE_00_MASTER to avoid duplicate names when multiple runs occur the same day.
    """
    now = datetime.now()
    return now.strftime("%Y%m%d_%H%M%S") + "_" + f"{now.microsecond // 1000:03d}"


def format_path(path: Path) -> str:
    """Format a path as a full absolute path string."""
    return str(path.resolve())


class DualOutput:
    """
    Capture output for terminal and Markdown file simultaneously.
    """

    def __init__(self, output_file: Path):
        self.output_file = output_file
        self._markdown_lines: list[str] = []

    def print(self, text: str, markdown_format: str | None = None, *, emoji: str | None = None) -> None:
        prefix = f"{emoji} " if emoji else ""
        line = f"{prefix}{text}"
        print(line)
        self._markdown_lines.append(markdown_format if markdown_format is not None else line)

    def save(self, header_info: dict[str, Any] | None = None) -> None:
        header_info = header_info or {}
        now = datetime.now()
        header = (
            "# Test Runner Output\n\n"
            f"**Library:** {header_info.get('library', '')}  \n"
            f"**Layer:** {header_info.get('layer', '')}  \n"
            f"**Generated:** {now.strftime('%d-%b-%Y %H:%M:%S')}.{now.microsecond // 1000:03d}  \n"
            f"**Runner:** {header_info.get('runner', 'TestRunner')}  \n\n"
            "---\n\n"
        )
        self.output_file.write_text(header + "\n".join(self._markdown_lines) + "\n", encoding="utf-8")


def print_header(title: str, output: DualOutput | None = None) -> None:
    sep = "=" * 80
    if output is None:
        print(sep)
        print(title)
        print(sep)
        return
    output.print(sep)
    output.print(title)
    output.print(sep)


def print_section(title: str, output: DualOutput | None = None) -> None:
    if output is None:
        print(f"\n{'=' * 80}\n{title}\n{'=' * 80}\n")
        return
    output.print(f"\n{title}", f"\n## {title}\n")


def print_status(success: bool, message: str, output: DualOutput | None = None) -> None:
    emoji = "✅" if success else "❌"
    if output is None:
        print(f"{emoji} {message}")
        return
    output.print(message, f"**Result:** {emoji} {message}", emoji=emoji)
@dataclass(frozen=True)

class PytestRunResult:
    exit_code: int


def run_pytest(
    *,
    test_dir: Path,
    markers: list[str] | None = None,
    extra_args: list[str] | None = None,
    cwd: Path | None = None,
) -> PytestRunResult:
    """
    Run pytest as a subprocess with standard eXonware flags.
    """
    markers = markers or []
    extra_args = extra_args or []
    cwd = cwd or test_dir
    args: list[str] = [
        sys.executable,
        "-m",
        "pytest",
        str(test_dir),
        "-v",
        "--tb=short",
        "-x",
        "--strict-markers",
    ]
    if markers:
        args.extend(["-m", " and ".join(markers)])
    args.extend(extra_args)
    completed = subprocess.run(args, cwd=str(cwd))
    return PytestRunResult(exit_code=int(completed.returncode))


class TestRunner:
    """
    Simple reusable test runner for a single directory/layer.
    """

    def __init__(
        self,
        *,
        library_name: str,
        layer_name: str,
        description: str,
        test_dir: Path,
        markers: list[str] | None = None,
        output_file: Path | None = None,
        pytest_cwd: Path | None = None,
    ):
        self.library_name = library_name
        self.layer_name = layer_name
        self.description = description
        self.test_dir = test_dir
        self.markers = markers or []
        # When set, pytest subprocess uses this cwd (e.g. project root for pytest.ini / src layout).
        self.pytest_cwd = pytest_cwd
        # Layer runners should pass output_file=None to not write files
        # Only main runner writes to docs/logs/tests/
        self.output_file = output_file
        if self.output_file is None:
            # Layer runner - create dummy output that won't save
            # Use a temporary path that we'll ignore
            import tempfile
            dummy_file = Path(tempfile.gettempdir()) / "xwsystem_test_runner_dummy.md"
            self.output = DualOutput(dummy_file)
            self._is_layer_runner = True
        else:
            self.output = DualOutput(self.output_file)
            self._is_layer_runner = False

    def run(self) -> int:
        ensure_utf8_console()
        print_header(self.description, self.output)
        self.output.print(
            f"Directory: {format_path(self.test_dir)}",
            f"**Directory:** `{format_path(self.test_dir)}`",
            emoji="📂",
        )
        cwd = self.pytest_cwd if self.pytest_cwd is not None else self.test_dir
        result = run_pytest(test_dir=self.test_dir, markers=self.markers, cwd=cwd)
        ok = result.exit_code == 0
        print_status(ok, "PASSED" if ok else "FAILED", self.output)
        # Layer runners don't write files - only main runner writes to docs/logs/tests/
        if not self._is_layer_runner:
            self.output.save(
                {
                    "library": self.library_name,
                    "layer": self.layer_name,
                    "runner": "TestRunner",
                }
            )
        return result.exit_code
